Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -1263,4 +1275,237 @@ lemma nonsingularLift_addMap {P Q : PointClass F} (hP : W.NonsingularLift P)
 
 end Addition
 
+section Point
+
+/-! ### Nonsingular rational points -/
+
+variable (W') in
+/-- A nonsingular rational point on `W'`. -/
+@[ext]
+structure Point where
+  {point : PointClass R}
+  (nonsingular : W'.NonsingularLift point)
+
+/-- The point class underlying a nonsingular rational point on `W'`. -/
+add_decl_doc Point.point
+
+/-- The nonsingular condition underlying a nonsingular rational point on `W'`. -/
+add_decl_doc Point.nonsingular
+
+namespace Point
+
+lemma mk_point {P : PointClass R} (h : W'.NonsingularLift P) : (mk h).point = P :=
+  rfl
+
+instance instZeroPoint [Nontrivial R] : Zero W'.Point :=
+  ⟨⟨nonsingularLift_zero⟩⟩
+
+lemma zero_def [Nontrivial R] : (0 : W'.Point) = ⟨nonsingularLift_zero⟩ :=
+  rfl
+
+lemma zero_point [Nontrivial R] : (0 : W'.Point).point = ⟦![1, 1, 0]⟧ :=
+  rfl
+
+/-- The map from a nonsingular rational point on a Weierstrass curve `W'` in affine coordinates
+to the corresponding nonsingular rational point on `W'` in Jacobian coordinates. -/
+def fromAffine [Nontrivial R] : W'.toAffine.Point → W'.Point
+  | 0 => 0
+  | Affine.Point.some h => ⟨(nonsingularLift_some ..).mpr h⟩
+
+lemma fromAffine_zero [Nontrivial R] : fromAffine 0 = (0 : W'.Point) :=
+  rfl
+
+lemma fromAffine_some [Nontrivial R] {X Y : R} (h : W'.toAffine.Nonsingular X Y) :
+    fromAffine (Affine.Point.some h) = ⟨(nonsingularLift_some ..).mpr h⟩ :=
+  rfl
+
+/-- The negation of a nonsingular rational point on `W`.
+Given a nonsingular rational point `P` on `W`, use `-P` instead of `neg P`. -/
+def neg (P : W.Point) : W.Point :=
+  ⟨nonsingularLift_negMap P.nonsingular⟩
+
+instance instNegPoint : Neg W.Point :=
+  ⟨neg⟩
+
+lemma neg_def (P : W.Point) : -P = P.neg :=
+  rfl
+
+lemma neg_point (P : W.Point) : (-P).point = W.negMap P.point :=
+  rfl
```

Comment:
for example, in my opinion these can be compressed to three consecutive lines.

Reply:
Even if we compress a few lemmas into shorter lines, in the long term there will be issues because we still have to add the analogous `map` lemmas after this, and likely many other things in the future. Note that I haven't added any `simps` to any of the lemmas here, so adding those might make it quite a bit longer too.

For now I will add `Jacobian.lean` into the style-exceptions. Once all these are merged I will split it into three files (TBD): API for triples and equations/nonsingularity (`Jacobian/Basic.lean` ~450 lines), group law polynomials and operations (`Jacobian/Operations.lean` ~950 lines), and the actual group law converted from the affine group law (`Jacobian/Group.lean` ~500 lines). The `map` lemmas will reside in whichever file the relevant definitions reside.

Once all that is done, I will mirror everything for affine and projective coordinates. Eventually we'll take all the `*_simp` stuff from all the files and put them into a standalone file in the same directory, so that will also reduce a few lines.

Reply:
I think `@[simp]` on the same line as def/lemma is even more common than removing blank lines between grouped lemmas. Searching for `@[simp] ` returns 3577 results across mathlib.
But it's certainly good if you can find a natural subdivision to split the file.